src/Confluent.Kafka/TopicPartitionOffset.cs

```diff
@@ -36,8 +48,24 @@ public class TopicPartitionOffset
         ///     The offset leader epoch (optional).
         /// </param>
         public TopicPartitionOffset(TopicPartition tp, Offset offset,
-                                    int? leaderEpoch = null)
```

Comment:
I don't know much about dotnet. What does this mean? `int? leaderEpoch = null`. 

If it was in Python, it would mean to take leaderEpoch as null if not provided. In that case the change wouldn't have caused any issue. I don't think this is the case here. Then what does this statement mean?

Reply:
`int?` means its nullable, ` = null` means null by default.

The issue stemmed from a transitive dependency.

Imagine L2 uses an older version of our library, which includes the original `TopicPartitionOffset` constructor. If an application A references the latest versions of both confluent-kafka-dotnet and L2, a `MissingMethodException` will be thrown if A calls an L2 method that relies on the old constructor, which was removed in the latest confluent-kafka-dotnet version.

To resolve this, we’re reintroducing the original constructor to ensure compatibility with libraries like L2 that haven't upgraded to the latest confluent-kafka-dotnet version.

Reply:
As discussed, this should have been working with optional parameter but its not working in transitive dependency.

Reply:
> I don't know much about dotnet. What does this mean? `int? leaderEpoch = null`.
> 
> If it was in Python, it would mean to take leaderEpoch as null if not provided. In that case the change wouldn't have caused any issue. I don't think this is the case here. Then what does this statement mean?

The difference between a constructor with an optional parameter and one without parameters is very clear in the intermediate code.
C# code:
```csharp
public class TestClass1
{
    public TestClass1(int? someValue = null)
    {
    }
}

public class TestClass2
{
    public TestClass2(int? someValue)
    {
    }

    public TestClass2() : this(null)
    {
    }
}

public static class TestClass
{
    public static void InitializeTestClass1InstanceWithoutParameters()
    {
        var instance = new TestClass1();
    }

    public static void InitializeTestClass2InstanceWithoutParameters()
    {
        var instance = new TestClass2();
    }
}
```

IL code:

![image](https://github.com/confluentinc/confluent-kafka-dotnet/assets/54648977/877e24fa-ce5c-47bb-8c41-adfece786c75)

In the case of optional parameters, their values are still passed to the constructor, but implicitly.

Moreover, the distinction between an optional parameter and no parameter allows the following set of constructors to be used without compilation errors (the default parameter value in this case is redundant: the constructor without an optional parameter takes precedence):
```csharp
public class TestClass1
{
    public TestClass1(int? someValue = null)
    {
    }

    public TestClass1()
    {
    }
}
```